### DISCUSSION TITLE
We may be using unoptimised shaders

### DISCUSSION THREAD

**Discussion by AnyOldName3** [vsg-dev/VulkanSceneGraph#1327] [General]
As a tangent to investigating something else, I realised that we're compiling shaders with `glslang::SpvOptions::disableOptimizer` set to `true`. That sounds bad, but at least historically, it doesn't seem to have been intended to optimise shaders properly anyway, and instead was just a badly-named variable controlling whether HLSL legalisation occurred when compiling a shader written in one language version to bytecode for a different shader model version.

The expectation, at least in 2019, was that the generated SPIR-V would be run through SPIRV-Tools-opt. I haven't been able to find anything confirming that that's changed.

If it is indeed still necessary, then it should be a fairly minor change to make, as we've already got a dependency on SPIRV-Tools via glslang.

**Comment by AnyOldName3** [comment]
I've pushed some work to address this here: https://github.com/AnyOldName3/VulkanSceneGraph/commit/e16430de4fec25f0c1ed993d62064edc4746805b

Here's what I wrote in the commit message:

> Unfortunately, on my conveniently-available hardware (AMD running Windows), it doesn't seem to make any difference.
> It's definitely changing the bytecode that's sent to the driver (I checked that thoroughly after an earlier revision of this work turned out to only actually be optimising the first shader it was given)
> 
> I don't know whether that's because:
> * AMD's Windows driver already optimises SPIR-V bytecode at least this well before running it. This is the most plausible explanation to me, though.
> * The VSG's stock shader isn't particularly amenable to optimisation.
> * I've not tested it on anything where shader execution (except unavoidable bits) takes a meaningful amount of time.
> 
> It would make a degree of sense for other people to test this on other machines - I'd be unsurprised if embedded platforms didn't bother optimising the SPIR-V they're given, but I don't have any such hardware myself.

The patch itself is relatively small except for the recompiled shaders (which can easily be recompiled again), so it should be fine to shelve this until it turns out to be necessary. Hopefully if anyone's hitting an optimisation-related problem on a non-desktop platform, this will show up in a search and they can try it without too much hassle.

**Comment by robertosfield** [comment]
I've done a first pass review and the changes look good, could you generate a PR from the changes?

**Reply by AnyOldName3** [reply]
https://github.com/vsg-dev/VulkanSceneGraph/pull/1381
